### PR TITLE
District dropdown

### PR DESCRIPTION
**Pull request by breville**
Builds upon work by @tanyaparker to have a district dropdown selector, used first in the workshop enrollment page.

### CODE REVIEW

File: dashboard/app/models/school_district.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  }
+
+  # Use the zero byte as the quote character to allow importing double quotes
+  #   via http://stackoverflow.com/questions/8073920/importing-csv-quoting-error-is-driving-me-nuts
+  CSV_IMPORT_OPTIONS = { col_sep: "\t", headers: true, :quote_char => "\x00" }
+
+  def self.find_or_create_all_from_tsv!(filename)
+    created = []
+    CSV.read(filename, CSV_IMPORT_OPTIONS).each do |row|
+      created << self.first_or_create_from_tsv_row!(row)
```
Comment (joshlory) [inline dashboard/app/models/school_district.rb:35]: Nit: prefer `map`.  Can this code be de-duped against the similar code in [`callout.rb`](https://github.com/code-dot-org/code-dot-org/blob/a7800a69e3153f7bb8d7bb32fe3e9c13f9170f40/dashboard/app/models/callout.rb#L37)?

File: dashboard/db/schema.rb [hunk 2]
```diff
@@ -555,6 +557,15 @@
   add_index "puzzle_ratings", ["script_id", "level_id"], name: "index_puzzle_ratings_on_script_id_and_level_id", using: :btree
   add_index "puzzle_ratings", ["user_id", "script_id", "level_id"], name: "index_puzzle_ratings_on_user_id_and_script_id_and_level_id", unique: true, using: :btree
 
+  create_table "school_districts", force: :cascade do |t|
+    t.string   "name",       limit: 255
```
Comment (ashercodeorg) [inline dashboard/db/schema.rb]: Should any or all of these be required (e.g., nonnull)?

File: dashboard/db/schema.rb [hunk 3]
```diff
@@ -555,6 +557,15 @@
...
+    t.string   "city",       limit: 255
+    t.string   "state",      limit: 255
+    t.string   "zip",        limit: 255
+    t.datetime "created_at",             null: false
+    t.datetime "updated_at",             null: false
```
Comment (ashercodeorg) [inline dashboard/db/schema.rb:576]: What do these timestamps represent?

File: dashboard/test/factories.rb [hunk 4]
```diff
@@ -1,4 +1,10 @@
-FactoryGirl.define do
+FactoryGirl.define do  factory :school_district do
+    name "MyString"
+city "MyString"
+state "MyString"
+zip "MyString"
```
Comment (ashercodeorg) [inline dashboard/test/factories.rb]: Please make these strings distinct to detect field mixups and make debugging easier.

File: dashboard/db/schema.rb [hunk 5]
```diff
@@ -918,6 +929,7 @@
   add_foreign_key "authored_hint_view_requests", "users"
   add_foreign_key "hint_view_requests", "users"
   add_foreign_key "level_concept_difficulties", "levels"
+  add_foreign_key "pd_enrollments", "school_districts"
```
Comment (ashercodeorg) [inline dashboard/db/schema.rb:943]: What happens when school districts change (no longer exist)?

File: dashboard/app/models/school_district.rb [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  include Seeded
+
+  CSV_HEADERS =
+  {
+    :id => 'LEAID',
+    :name => 'NAME',
+    :city => 'LCITY',
+    :state => 'LSTATE',
+    :zip => 'LZIP'
+  }
```
Comment (ashercodeorg) [inline dashboard/app/models/school_district.rb:26]: Where are the strings emanating from? A comment would be nice explaining the (my presumed) dependency on some school district listing?

File: dashboard/db/schema.rb (hunk 3 above)
Comment (breville) [inline dashboard/db/schema.rb:576]: Oh, I guess creating a migration automatically created these.  They are ActiveRecord features, and seem to feature in most of our models:

http://guides.rubyonrails.org/active_record_basics.html#schema-conventions

File: dashboard/db/schema.rb (hunk 5 above)
Comment (breville) [inline dashboard/db/schema.rb:943]: Good question.  I like having a foreign key constraint just to maintain some data integrity.  But perhaps we could add an additional "deleted" column to school_districts if/when we start removing districts?

File: pegasus/sites.v3/code.org/public/educate/professional-learning/exploring-cs-apply.md [hunk 7]
```diff
@@ -10,10 +10,7 @@ We welcome teachers to sign up for the Exploring Computer Science (ECS) Professi
 
 The ECS professional learning program is open to teachers in partner districts who are interested in teaching the course. Check to see if your district has partnered to make ECS available in your region for the 2016-2017 school year. If you live in Idaho, Indiana, Utah, or Virginia, your district is covered by a statewide partnership. Click on the icon in your state to apply.
 
-<br/> 
-<br/>
-<iframe src="https://www.google.com/maps/d/u/0/embed?mid=z3jxShb6X1IM.khZ1NwyRM_3Y" width="640" height="480"></iframe>
-
+<%= view :district_dropdown %>
```
Comment (breville) [inline pegasus/sites.v3/code.org/public/educate/professional-learning/exploring-cs-apply.md]: @tanyaparker is this page live yet?  if so, i should just leave the map in place for now.

Comment (tanyaparker) [inline pegasus/sites.v3/code.org/public/educate/professional-learning/exploring-cs-apply.md]: Yep, this page is live and we should leave in the map. For spectators, we shouldn't change it to the dropdown until we save the info in a pegasus form and know where to redirect each district (application, waitlist, exception).

File: dashboard/app/models/school_district.rb (hunk 1 above)
Comment (aoby) [inline dashboard/app/models/school_district.rb:35]: also nit: prefer newer ruby hash syntax (no unnecessary hash rocket)

``` ruby
{
  id: 'LEAID',
  name: 'NAME',
  ...
```

File: dashboard/app/models/school_district.rb [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+  def self.find_or_create_all_from_tsv!(filename)
+    created = []
+    CSV.read(filename, CSV_IMPORT_OPTIONS).each do |row|
+      created << self.first_or_create_from_tsv_row!(row)
+    end
+    created
+  end
+
+  def self.first_or_create_from_tsv_row!(row_data)
+    params = {id: row_data[CSV_HEADERS[:id]],
```
Comment (aoby) [inline dashboard/app/models/school_district.rb]: nit, start items on a new line for cleaner indentation:

``` ruby
params = {
  id: row_data[...],
  name: ...
  ...
}
```

File: dashboard/config/initializers/active_model_serializers.rb [hunk 9]
```diff
@@ -12,20 +12,4 @@ def attributes(obj)
 end
 class ActiveModel::ErrorsSerializer < HashSerializer; end
 
-# Patch to provide more informative missing-serializer error message
```
Comment (aoby) [inline dashboard/config/initializers/active_model_serializers.rb]: +1 This same change is in my PR (with some context in the description): https://github.com/code-dot-org/code-dot-org/pull/8451

File: dashboard/db/schema.rb (hunk 5 above)
Comment (aoby) [inline dashboard/db/schema.rb:943]: `deleted_at` column

File: dashboard/lib/tasks/seed.rake [hunk 10]
```diff
@@ -111,6 +111,13 @@ namespace :seed do
     end
   end
 
+  task school_districts: :environment do
+    SchoolDistrict.transaction do
+      #SchoolDistrict.reset_db
```
Comment (aoby) [inline dashboard/lib/tasks/seed.rake]: ?

File: dashboard/test/models/school_district_test.rb [hunk 11]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class SchoolDistrictTest < ActiveSupport::TestCase
```
Comment (aoby) [inline dashboard/test/models/school_district_test.rb]: Either remove this empty test class, or better fill it in with tests ;)

File: pegasus/sites.v3/code.org/public/educate/professional-learning/exploring-cs-apply.md [hunk 12]
```diff
@@ -3,14 +3,14 @@ title: Find ECS near you
 nav: plc_ecs_nav
 ---
 # Find the Exploring Computer Science program near you
-We welcome teachers to sign up for the Exploring Computer Science (ECS) Professional Learning Program. The ECS program supports teachers with diverse teaching backgrounds. Whether you are new to teaching CS or you have CS teaching experience outside this course, the program is designed to promote growth in your teaching practice, providing space for teachers to become comfortable with the curricular materials and associated teaching strategies. 
+We welcome teachers to sign up for the Exploring Computer Science (ECS) Professional Learning Program. The ECS program supports teachers with diverse teaching backgrounds. Whether you are new to teaching CS or you have CS teaching experience outside this course, the program is designed to promote growth in your teaching practice, providing space for teachers to become comfortable with the curricular materials and associated teaching strategies.
```
Comment (aoby) [inline pegasus/sites.v3/code.org/public/educate/professional-learning/exploring-cs-apply.md:6]: Is this related?

File: dashboard/test/models/school_district_test.rb (hunk 11 above)
Comment (breville) [inline dashboard/test/models/school_district_test.rb]: Testing this particular model is a bit boring.. testing the controller that provides the API is a little more interesting, especially if the database is fully seeded before unit tests (is it?)

File: pegasus/sites.v3/code.org/public/educate/professional-learning/exploring-cs-apply.md (hunk 12 above)
Comment (breville) [inline pegasus/sites.v3/code.org/public/educate/professional-learning/exploring-cs-apply.md:6]: I touched it because it's another destination for district dropdown :)
